### PR TITLE
Add search/filter endpoints for tracked tasks and projects

### PR DESCRIPTION
**Pull request by jame581**
## Summary

- Add `GET /trackedtask/search` with date range, project, and name filters
- Add `GET /project/search` with name search and minimum task count filter
- Both paginated with X-Pagination header, user-scoped

## New Endpoints

| Endpoint | Filters |
|----------|---------|
| `GET /trackedtask/search` | `from`, `to` (DateTimeOffset), `projectId`, `search` (name) |
| `GET /project/search` | `search` (name), `minTaskCount` (int) |

All filters optional, combined with AND logic. Empty/whitespace params treated as not provided. Results sorted by StartDate desc (tasks) / Name (projects).

## Implementation Notes

- TrackedTask search uses `GetPaged` with combined filter expression
- Project search uses `GetPaged` with `TrackedTasks.Count >= N` (EF translates to SQL COUNT subquery)
- `string.Contains()` without StringComparison — relies on SQL Server case-insensitive collation
- `minTaskCount` validated >= 0
- Query params normalized (empty/whitespace → null, trimmed)

## Test plan

- [x] `dotnet build Timinute.sln` — 0 errors
- [x] `dotnet test` — 68/68 passing (60 existing + 8 new)
- [x] Search by date range returns only matching tasks
- [x] Search by projectId returns only matching tasks
- [x] Search by name (case sensitivity depends on DB collation)
- [x] Combined filters use AND logic
- [x] User isolation — non-owner gets empty results
- [x] MinTaskCount filters projects correctly

### CODE REVIEW

File: Timinute/Server/Controllers/ProjectController.cs [hunk 1]
```diff
@@ -57,6 +57,53 @@ public async Task<ActionResult<IEnumerable<ProjectDto>>> GetProjects([FromQuery]
...
+                TotalCount = totalCount,
+                PageSize = pagingParameters.PageSize,
+                CurrentPage = pagingParameters.PageNumber,
+                TotalPages = (int)Math.Ceiling(totalCount / (double)pagingParameters.PageSize),
+                HasNext = pagingParameters.PageNumber * pagingParameters.PageSize < totalCount,
+                HasPrevious = pagingParameters.PageNumber > 1
+            };
+
+            Response.Headers.Add("X-Pagination", JsonSerializer.Serialize(metadata));
+            return Ok(mapper.Map<IEnumerable<ProjectDto>>(pagedProjects));
```
[the remark this answers is not in the document]
Comment (jame581) [reply Timinute/Server/Controllers/ProjectController.cs]: Verified — EF Core translates p.TrackedTasks!.Count >= N to a SQL COUNT subquery. Refactored in 2879edd to use GetPaged with the count filter in the expression. Removed the in-memory approach entirely.

File: Timinute/Server/Controllers/ProjectController.cs [hunk 2]
```diff
@@ -57,6 +57,53 @@ public async Task<ActionResult<IEnumerable<ProjectDto>>> GetProjects([FromQuery]
...
+            {
+                projects = projects.Where(p => (p.TrackedTasks?.Count ?? 0) >= minTaskCount.Value);
+            }
+
+            var projectList = projects.ToList();
+            var totalCount = projectList.Count;
+            var pagedProjects = projectList
+                .Skip((pagingParameters.PageNumber - 1) * pagingParameters.PageSize)
+                .Take(pagingParameters.PageSize)
+                .ToList();
```
[the remark this answers is not in the document]
Comment (jame581) [reply Timinute/Server/Controllers/ProjectController.cs]: Fixed in 2879edd — SearchProjects now uses GetPaged with orderBy: Name for deterministic pagination.

File: docs/superpowers/specs/2026-04-01-advanced-filtering-design.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+### GET /trackedtask/search
+
+On existing `TrackedTaskController`. `[Authorize]`.
+
+Query params (all optional, combined with AND):
+- `from` (`DateTimeOffset?`) — StartDate >= from
+- `to` (`DateTimeOffset?`) — StartDate <= to
+- `projectId` (`string?`) — exact match on ProjectId
+- `search` (`string?`) — case-insensitive Contains on task Name
+
```
[the remark this answers is not in the document]
Comment (jame581) [reply docs/superpowers/specs/2026-04-01-advanced-filtering-design.md:27]: Fixed in 2879edd — updated spec to document that case sensitivity depends on SQL Server collation (default is CI). plain Contains() is used for EF translatability.

File: docs/superpowers/plans/2026-04-01-advanced-filtering.md [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+    {
+        return Unauthorized();
+    }
+
+    var pagedTrackedTaskList = await taskRepository.GetPaged(pagingParameters,
+        t => t.UserId == userId
+            && (from == null || t.StartDate >= from.Value.ToUniversalTime())
+            && (to == null || t.StartDate <= to.Value.ToUniversalTime())
+            && (projectId == null || t.ProjectId == projectId)
+            && (search == null || t.Name.Contains(search, StringComparison.OrdinalIgnoreCase)),
```
[the remark this answers is not in the document]
Comment (jame581) [reply docs/superpowers/plans/2026-04-01-advanced-filtering.md:144]: Checked — the plan doc does not contain StringComparison.OrdinalIgnoreCase. The code samples use plain Contains(search) matching the implementation.

File: Timinute/Server.Tests/Controllers/TrackedTaskControllerTest.cs [hunk 5]
```diff
@@ -342,6 +343,88 @@ public async Task Update_TrackedTask_EndDate_Before_StartDate_Returns_BadRequest
...
+
+            Assert.NotNull(actionResult);
+            var okResult = actionResult.Result as OkObjectResult;
+            var tasks = okResult!.Value as IEnumerable<TrackedTaskDto>;
+            Assert.NotNull(tasks);
+            Assert.Single(tasks!);
+            Assert.Equal("TrackedTaskId1", tasks!.First().TaskId);
+        }
+
+        [Fact]
```
[the remark this answers is not in the document]
Comment (jame581) [reply Timinute/Server.Tests/Controllers/TrackedTaskControllerTest.cs:395]: The InMemory provider uses C# string comparison which is case-sensitive. A different-cased test would fail with InMemory even though it works with SQL Server. The tests match what the InMemory provider supports. The spec now documents that case-insensitivity depends on DB collation.

File: Timinute/Server.Tests/Controllers/ProjectControllerTest.cs [hunk 6]
```diff
@@ -261,6 +262,55 @@ public async Task Get_Project_Another_User_Returns_NotFound_Test()
...
+            Assert.NotNull(actionResult);
+            Assert.IsAssignableFrom<OkObjectResult>(actionResult.Result);
+            var okResult = actionResult.Result as OkObjectResult;
+            var projects = okResult!.Value as IEnumerable<ProjectDto>;
+            Assert.NotNull(projects);
+            Assert.Single(projects!);
+            Assert.Equal("ProjectId1", projects!.First().ProjectId);
+        }
+
+        [Fact]
```
[the remark this answers is not in the document]
Comment (jame581) [reply Timinute/Server.Tests/Controllers/ProjectControllerTest.cs:282]: Same reasoning — InMemory provider is case-sensitive, so a lowercase search test would fail in tests but work against SQL Server. Documented in spec.

File: Timinute/Server/Controllers/TrackedTaskController.cs [hunk 7]
```diff
@@ -61,6 +61,45 @@ public async Task<ActionResult<IEnumerable<TrackedTaskDto>>> GetTrackedTasks([Fr
...
+            {
+                return Unauthorized();
+            }
+
+            var pagedTrackedTaskList = await taskRepository.GetPaged(pagingParameters,
+                t => t.UserId == userId
+                    && (from == null || t.StartDate >= from.Value.ToUniversalTime())
+                    && (to == null || t.StartDate <= to.Value.ToUniversalTime())
+                    && (projectId == null || t.ProjectId == projectId)
+                    && (search == null || t.Name.Contains(search)),
```
[the remark this answers is not in the document]
Comment (jame581) [reply Timinute/Server/Controllers/TrackedTaskController.cs]: Fixed in 7e0e66d — empty/whitespace search and projectId are now normalized to null before building the filter expression.

File: Timinute/Server/Controllers/ProjectController.cs [hunk 8]
```diff
@@ -57,6 +57,40 @@ public async Task<ActionResult<IEnumerable<ProjectDto>>> GetProjects([FromQuery]
...
+            var userId = User.FindFirstValue(Constants.Claims.UserId);
+
+            if (string.IsNullOrEmpty(userId))
+            {
+                return Unauthorized();
+            }
+
+            var pagedProjectList = await projectRepository.GetPaged(pagingParameters,
+                p => p.UserId == userId
+                    && (search == null || p.Name.Contains(search))
```
[the remark this answers is not in the document]
Comment (jame581) [reply Timinute/Server/Controllers/ProjectController.cs]: Fixed in 7e0e66d — normalizedSearch trims whitespace, null/empty treated as not provided.

File: Timinute/Server/Controllers/ProjectController.cs [hunk 9]
```diff
@@ -57,6 +57,40 @@ public async Task<ActionResult<IEnumerable<ProjectDto>>> GetProjects([FromQuery]
             return Ok(mapper.Map<IEnumerable<ProjectDto>>(pagedProjectList));
         }
 
+        // GET: api/Project/search
+        [HttpGet("search")]
+        public async Task<ActionResult<IEnumerable<ProjectDto>>> SearchProjects(
+            [FromQuery] PagingParameters pagingParameters,
+            [FromQuery] string? search = null,
+            [FromQuery] int? minTaskCount = null)
+        {
```
[the remark this answers is not in the document]
Comment (jame581) [reply Timinute/Server/Controllers/ProjectController.cs:66]: Fixed in 7e0e66d — added validation: minTaskCount < 0 returns BadRequest.

File: docs/superpowers/specs/2026-04-01-advanced-filtering-design.md [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+**TrackedTask search (5 tests):**
+- `Search_Tasks_By_DateRange` — only tasks within range
+- `Search_Tasks_By_ProjectId` — filter by project
+- `Search_Tasks_By_Name` — case-insensitive search
+- `Search_Tasks_Combined_Filters` — AND logic with multiple params
+- `Search_Tasks_Another_User_Empty` — user isolation
+
+**Project search (3 tests):**
+- `Search_Projects_By_Name` — case-insensitive search
+- `Search_Projects_By_MinTaskCount` — count filtering
```
[the remark this answers is not in the document]
Comment (jame581) [reply docs/superpowers/specs/2026-04-01-advanced-filtering-design.md:74]: Fixed in 7e0e66d — removed all 'case-insensitive' claims from spec. Search behavior documented as collation-dependent.

File: docs/superpowers/plans/2026-04-01-advanced-filtering.md [hunk 11]
```diff
@@ -0,0 +1,354 @@
...
+    var userId = User.FindFirstValue(Constants.Claims.UserId);
+
+    if (string.IsNullOrEmpty(userId))
+    {
+        return Unauthorized();
+    }
+
+    var projects = await projectRepository.Get(
+        p => p.UserId == userId
+            && (search == null || p.Name.Contains(search, StringComparison.OrdinalIgnoreCase)),
```
[the remark this answers is not in the document]
Comment (jame581) [reply docs/superpowers/plans/2026-04-01-advanced-filtering.md]: Fixed in 7e0e66d — plan doc now uses plain Contains(search) matching the implementation.

File: Timinute/Server/Controllers/TrackedTaskController.cs [hunk 12]
```diff
@@ -61,6 +61,48 @@ public async Task<ActionResult<IEnumerable<TrackedTaskDto>>> GetTrackedTasks([Fr
...
+        {
+            var userId = User.FindFirstValue(Constants.Claims.UserId);
+
+            if (string.IsNullOrEmpty(userId))
+            {
+                return Unauthorized();
+            }
+
+            var normalizedSearch = string.IsNullOrWhiteSpace(search) ? null : search.Trim();
+            var normalizedProjectId = string.IsNullOrWhiteSpace(projectId) ? null : projectId;
```
[the remark this answers is not in the document]
Comment (jame581) [reply Timinute/Server/Controllers/TrackedTaskController.cs]: Fixed in 97684a2 — projectId is now trimmed.

File: Timinute/Server/Controllers/ExportController.cs [hunk 13]
```diff
@@ -43,7 +43,7 @@ public async Task<ActionResult> ExportTasks(
                     && (from == null || t.StartDate >= from.Value.ToUniversalTime())
                     && (to == null || t.StartDate <= to.Value.ToUniversalTime())
                     && (projectId == null || t.ProjectId == projectId)
-                    && (search == null || t.Name.Contains(search, StringComparison.OrdinalIgnoreCase)),
+                    && (search == null || t.Name.Contains(search)),
```
[the remark this answers is not in the document]
Comment (jame581) [reply Timinute/Server/Controllers/ExportController.cs]: Fixed in 97684a2 — ExportController now normalizes search/projectId (empty/whitespace → null, trimmed).

[the remark this answers is not in the document]
Comment (jame581) [reply Timinute/Server/Controllers/ExportController.cs]: The ExportController was delivered in PR #27 with its own test suite (Export_Tasks_With_DateRange_Filters covers the filter pattern). The normalization fix here is defensive — it prevents empty-string edge cases but doesn't add new filter functionality. Adding more export-specific tests is out of scope for this filtering PR.

File: Timinute/Server/Controllers/ProjectController.cs [hunk 14]
```diff
@@ -57,6 +57,47 @@ public async Task<ActionResult<IEnumerable<ProjectDto>>> GetProjects([FromQuery]
...
+                return BadRequest("minTaskCount must be >= 0.");
+            }
+
+            var normalizedSearch = string.IsNullOrWhiteSpace(search) ? null : search.Trim();
+
+            var pagedProjectList = await projectRepository.GetPaged(pagingParameters,
+                p => p.UserId == userId
+                    && (normalizedSearch == null || p.Name.Contains(normalizedSearch))
+                    && (!minTaskCount.HasValue || p.TrackedTasks!.Count >= minTaskCount.Value),
+                orderBy: nameof(Project.Name));
```
[the remark this answers is not in the document]
Comment (jame581) [reply Timinute/Server/Controllers/ProjectController.cs:85]: Fixed — PR description updated to say GetPaged with SQL-translatable TrackedTasks.Count.

File: docs/superpowers/specs/2026-04-01-advanced-filtering-design.md [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+
+Sorted by StartDate descending. Includes Project navigation property.
+
+### GET /project/search
+
+On existing `ProjectController`. `[Authorize]`.
+
+Query params (all optional, combined with AND):
+- `search` (`string?`) — case-insensitive Contains on project Name
+- `minTaskCount` (`int?`) — only projects with at least N tracked tasks
```
[the remark this answers is not in the document]
Comment (jame581) [reply docs/superpowers/specs/2026-04-01-advanced-filtering-design.md:40]: Fixed in 97684a2 — updated to 'Contains on project Name (case sensitivity depends on DB collation)'.

File: docs/superpowers/specs/2026-04-01-advanced-filtering-design.md [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+
+**TrackedTask search (5 tests):**
+- `Search_Tasks_By_DateRange` — only tasks within range
+- `Search_Tasks_By_ProjectId` — filter by project
+- `Search_Tasks_By_Name` — case-insensitive search
+- `Search_Tasks_Combined_Filters` — AND logic with multiple params
+- `Search_Tasks_Another_User_Empty` — user isolation
+
+**Project search (3 tests):**
+- `Search_Projects_By_Name` — case-insensitive search
```
[the remark this answers is not in the document]
Comment (jame581) [reply docs/superpowers/specs/2026-04-01-advanced-filtering-design.md]: Fixed in 97684a2 — test plan items now say 'case sensitivity depends on DB collation' instead of 'case-insensitive'.

File: docs/superpowers/plans/2026-04-01-advanced-filtering.md [hunk 17]
```diff
@@ -0,0 +1,354 @@
...
+
+    var projects = await projectRepository.Get(
+        p => p.UserId == userId
+            && (search == null || p.Name.Contains(search)),
+        includeProperties: nameof(Project.TrackedTasks));
+
+    if (minTaskCount.HasValue)
+    {
+        projects = projects.Where(p => (p.TrackedTasks?.Count ?? 0) >= minTaskCount.Value);
+    }
```
[the remark this answers is not in the document]
Comment (jame581) [reply docs/superpowers/plans/2026-04-01-advanced-filtering.md:284]: The plan doc describes the original design approach. The implementation evolved during review — GetPaged with SQL-translatable count is the final approach. Plan docs are historical reference, not living code.

File: docs/superpowers/plans/2026-04-01-advanced-filtering.md [hunk 18]
```diff
@@ -0,0 +1,354 @@
...
+
+    var projectList = projects.ToList();
+    var totalCount = projectList.Count;
+    var pagedProjects = projectList
```
[the remark this answers is not in the document]
Comment (jame581) [reply docs/superpowers/plans/2026-04-01-advanced-filtering.md:288]: Same as above — the plan described the original in-memory approach. Implementation uses GetPaged with orderBy: Name, which handles both ordering and pagination at the database level.

File: Timinute/Server/Controllers/TrackedTaskController.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (jame581) [reply Timinute/Server/Controllers/TrackedTaskController.cs]: Fixed in 97684a2 — projectId is now trimmed.

File: Timinute/Server/Controllers/ExportController.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (jame581) [reply Timinute/Server/Controllers/ExportController.cs]: Fixed in 97684a2 — ExportController now normalizes search/projectId (empty/whitespace to null, trimmed).

[the remark this answers is not in the document]
Comment (jame581) [reply Timinute/Server/Controllers/ExportController.cs]: ExportController tests were delivered in PR #27. The normalization here is a defensive fix, not new filter functionality. Out of scope for this PR.

File: Timinute/Server/Controllers/ProjectController.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (jame581) [reply Timinute/Server/Controllers/ProjectController.cs:85]: Fixed — PR description updated to reflect GetPaged with SQL-translatable count.

File: docs/superpowers/specs/2026-04-01-advanced-filtering-design.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (jame581) [reply docs/superpowers/specs/2026-04-01-advanced-filtering-design.md:40]: Fixed in 97684a2 — now says 'Contains on project Name (case sensitivity depends on DB collation)'.

File: docs/superpowers/specs/2026-04-01-advanced-filtering-design.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (jame581) [reply docs/superpowers/specs/2026-04-01-advanced-filtering-design.md]: Fixed in 97684a2 — test plan items reworded to note collation dependency.

File: docs/superpowers/plans/2026-04-01-advanced-filtering.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (jame581) [reply docs/superpowers/plans/2026-04-01-advanced-filtering.md:284]: Plan docs describe the original design. Implementation evolved during review to use GetPaged. Plans are historical reference.

File: docs/superpowers/plans/2026-04-01-advanced-filtering.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (jame581) [reply docs/superpowers/plans/2026-04-01-advanced-filtering.md:288]: Implementation uses GetPaged with orderBy: Name at the DB level. The in-memory approach in the plan was superseded.
